python/pyspark/sql/connect/client/reattach.py

```diff
@@ -0,0 +1,228 @@
+    ExecutePlanResponse on the iterator to return a new iterator from server that continues after
+    that.
+
+    Since in reattachable execute the server does buffer some responses in case the client needs to
+    backtrack
+    """
+
+    def __init__(
+        self,
+        request: pb2.ExecutePlanRequest,
+        stub: grpc_lib.SparkConnectServiceStub,
+        retry_policy: Dict[str, Any],
+    ):
+        self._request = request
+        self._retry_policy = retry_policy
+        if request.operation_id:
+            self._operation_id = request.operation_id
+        else:
+            # Add operation id, if not present.
+            # with operationId set by the client, the client can use it to try to reattach on error
+            # even before getting the first response. If the operation in fact didn't even reach the
+            # server, that will end with INVALID_HANDLE.OPERATION_NOT_FOUND error.
+            self._operation_id = str(uuid.uuid4())
+
+        self._stub = stub
+        request.request_options.append(
+            pb2.ExecutePlanRequest.RequestOption(
+                reattach_options=pb2.ReattachOptions(reattachable=True)
+            )
+        )
+        self._initial_request = request
+
+        # ResponseId of the last response returned by next()
+        self._last_returned_response_id: Optional[str] = None
+
+        # True after ResponseComplete message was seen in the stream.
+        # Server will always send this message at the end of the stream, if the underlying iterator
+        # finishes without producing one, another iterator needs to be reattached.
+        self._result_complete = False
+
+        # Initial iterator comes from ExecutePlan request.
+        self._iterator: Iterator[pb2.ExecutePlanResponse] = self._stub.ExecutePlan(
+            self._initial_request
+        )
+
+        # Current item from this iterator.
+        self._current: Optional[pb2.ExecutePlanResponse] = None
+
+    def send(self, value: Any) -> pb2.ExecutePlanResponse:
+        # will trigger reattach in case the stream completed without result_complete
+        if not self._has_next():
+            raise StopIteration()
+
+        ret = self._current
+        assert ret is not None
+
+        self._last_returned_response_id = ret.response_id
+        if ret.result_complete:
+            self._result_complete = True
+            self._release_execute(None)  # release all
+        else:
+            self._release_execute(self._last_returned_response_id)
+        self._current = None
+        return ret
+
+    def _has_next(self) -> bool:
+        from pyspark.sql.connect.client.core import SparkConnectClient
+        from pyspark.sql.connect.client.core import Retrying
+
+        if self._result_complete:
+            # After response complete response
+            return False
+        else:
+            first_try = True
+            for attempt in Retrying(
```

Comment:
attempt is of type `AttemptManager` which contains a `_retry_state` that has a _count. We can make the retry state visible in the attempt manager if you want to avoid having a first try variable

Reply:
Yup, this is the time that this context manager shines :-)